test/test_plugins.py

```diff
@@ -401,6 +402,155 @@ def dummy9(self, **kwargs):
+                        ui.commands.PromptChoice('r', 'baR', None)]
+
+        self.register_plugin(DummyPlugin)
+        # Default options + extra choices by the plugin ('Foo', 'Bar')
+        opts = (u'Apply', u'More candidates', u'Skip', u'Use as-is',
+                u'Enter search',
+                u'enter Id', u'aBort') + ('Foo', 'baR')
+
+        config['import']['singletons'] = True
+        self.importer.add_choice(action.SKIP)
+        self.importer.run()
+        self.mock_input_options.assert_called_with(opts, default='a',
+                                                   require=ANY)
+
+    def test_choices_conflicts(self):
+        """Test the short letter conflict solving."""
+        class DummyPlugin(plugins.BeetsPlugin):
+            def __init__(self):
+                super(DummyPlugin, self).__init__()
+                self.register_listener('before_choose_candidate',
+                                       self.return_choices)
+
+            def return_choices(self, session, task):
+                return [ui.commands.PromptChoice('a', 'A foo', None),  # dupe
+                        ui.commands.PromptChoice('z', 'baZ', None),    # ok
+                        ui.commands.PromptChoice('z', 'Zupe', None),   # dupe
+                        ui.commands.PromptChoice('z', 'Zoo', None)]    # dupe
+
+        self.register_plugin(DummyPlugin)
+        # Default options + not dupe extra choices by the plugin ('baZ')
+        opts = (u'Apply', u'More candidates', u'Skip', u'Use as-is',
+                u'as Tracks', u'Group albums', u'Enter search',
+                u'enter Id', u'aBort') + ('baZ',)
+        self.importer.add_choice(action.SKIP)
+        self.importer.run()
+        self.mock_input_options.assert_called_once_with(opts, default='a',
+                                                        require=ANY)
+
+    def test_plugin_callback(self):
+        """Test that plugin callbacks are being called upon user choice."""
+        class DummyPlugin(plugins.BeetsPlugin):
+            def __init__(self):
+                super(DummyPlugin, self).__init__()
+                self.register_listener('before_choose_candidate',
+                                       self.return_choices)
+
+            def return_choices(self, session, task):
+                return [ui.commands.PromptChoice('f', 'Foo', self.foo)]
+
+            def foo(self, session, task):
+                pass
+
+        self.register_plugin(DummyPlugin)
+        # Default options + extra choices by the plugin ('Foo', 'Bar')
+        opts = (u'Apply', u'More candidates', u'Skip', u'Use as-is',
+                u'as Tracks', u'Group albums', u'Enter search',
+                u'enter Id', u'aBort') + ('Foo',)
+
+        # DummyPlugin.foo() should be called once
+        with patch.object(DummyPlugin, 'foo', autospec=True) as mock_foo:
+            with helper.control_stdin('\n'.join(['f', 's'])):
```

Comment:
This is a bit hackish, but `TestTerminalImportSession._add_choice_input` has some checks in place that prevent the use of `self.importer.add_choice('f')` directly.